HIP/hip-1056.md

```diff
@@ -0,0 +1,1769 @@
+processing.
+
+The division into block items also enables filtering of data based on data type. For example a downstream services could
+filter for only transactional data if there is no requirement for event or state information. Another service could
+filter on only accounts they are interested in. Block items removed from a filtered stream are represented by one or
+more `FilteredItemHash`’s which provide the missing hashes required to validate the filtered stream with the same block
+proof as the full stream.
+
+#### BlockHeader
+
+### BlockHeader
+
+A `BlockHeader` is one type of `output` BlockItem and will always be the first element in a Block. It contains Block
+metadata which describes a block’s identifying information such as its number and HAPI specification version
+
+```protobuf
+message BlockHeader {
+    /**
+     * Version of the HAPI specification that was used to serialize the block.
+     */
+    proto.SemanticVersion hapi_proto_version = 1;
+
+    /**
+     * The software version used to execute transactions in this block.
+     */
+    proto.SemanticVersion software_version = 2;
+
+    /**
+     * The block number of this block.
+     * <p>
+     * This value MUST be exactly `1` more than the previous block.<br/>
+     * Client systems SHOULD optimistically reject any block with a gap or
+     * reverse in `number` sequence, and MAY assume the block stream has
+     * encountered data loss, data corruption, or unauthorized modification.
+     */
+    uint64 number = 3;
+
+    /**
+     * The block hash of the previous block.
+     * This is here so a client parsing the BlockHeader can optimistically reject
+     * a block if this hash does not match the root hash of the previous block.
+     */
+    bytes previous_block_hash = 4;
+
+    /**
+     * A consensus timestamp for the start of this block.
+     * <p>
+     * This SHALL be the timestamp assigned by the hashgraph consensus
+     * algorithm to the first transaction of this block.
+     */
+    proto.Timestamp first_transaction_consensus_time = 5;
+
+    /**
+     * The hash algorithm used in this block.
+     */
+    proto.BlockHashAlgorithm hash_algorithm = 6;
+
+    /**
+     * Version of the active address book of the network at the time relating to the network public key
+     */
+    proto.SemanticVersion address_book_version = 7;
+}
+
+// This definition is in basic_types.proto.
+/**
+ * A specific hash algorithm used within a block.
+ *
+ * We did not reuse HashAlgorithm here because in all cases for now it will be SHA2_384 and if that's the default value
+ * then we can save space by not serializing it, whereas HASH_ALGORITHM_UNKNOWN = 0; is the default for HashAlgorithm.
+ */
+enum BlockHashAlgorithm {
+    SHA2_384 = 0;
+}
+```
+
+<aside>
+🚨 **Open Detail:** How do we define the `first_transaction_consensus_time` of a block with no transactions? Can it be
+first event consensus time if no transactions? what about empty block with no events?
```

Comment:
Some details and edge cases:
1. For "normal" rounds that contain events and transactions, the round timestamp is equal to the timestamp of the last transaction in the round
2. For empty rounds (no events), the round timestamp is equal to the timestamp of the last transaction in the previous round plus 1,000 nanos, rounded up to the nearest 1,000 nanos. Is there is no previous round, then the round timestamp is the median of the judge created times.

To me, #1 makes no sense at all. An event's consensus timestamp is equal to the timestamp of the FIRST transaction it contains. If the round follow the same rules, the round's timestamp would be equal to the timestamp of the first event and transaction it contains. As far as I'm aware, there is no reason we cannot change the round timestamp to work this way if we want. If we did that, we could just put the round timestamp in this protobuf instead of the first transaction timestamp and call it a day.

If we can't do that or don't want to do that, then we have some other options:
1. Keep the transaction timestamp but make it nullable. If there are no transactions in the round, it is null.
2. Keep the timestamp and make it always populated (with the one exception of an empty block sent after the freeze round), but for rounds with no transactions, it will be the round timestamp
3. Add a second field for the round timestamp which will always be populated (with the one exception of an empty block sent after the freeze round).

None of these sound good because there are exceptions and edge cases that need to be explained.